core/amber/src/main/scala/edu/uci/ics/texera/web/resource/ResultDownloadResource.scala

```diff
@@ -0,0 +1,202 @@
+  ): ResultDownloadResponse = {
+    // look for the response cached in the session to avoid creating duplicated file
+    if (
+      session.getUserProperties.containsKey(
+        getSessionName(request.downloadType)
+      )
+    ) {
+      return session.getUserProperties
+        .get(getSessionName(request.downloadType))
+        .asInstanceOf[ResultDownloadResponse]
+    }
+
+    // By now the workflow should finish running. Only one operator should contain results.
+    val count = sessionResults.count(p => !p._2.isEmpty)
+    if (count == 0) {
+      return ResultDownloadResponse(
+        request.downloadType,
+        "",
+        "The workflow contains no results"
+      )
+    } else if (count > 1) {
+      return ResultDownloadResponse(
+        request.downloadType,
+        "",
+        "The workflow does not finish running"
+      )
+    }
+    val result: List[ITuple] =
+      sessionResults.map(p => p._2).find(p => !p.isEmpty).get
+
+    var response: ResultDownloadResponse = null
+    request.downloadType match {
+      case "google_sheet" =>
+        response = handleGoogleSheetRequest(request, result)
+      case _ =>
+        response = ResultDownloadResponse(
+          request.downloadType,
+          "",
+          s"Unknown download type: ${request.downloadType}"
+        )
+    }
+
+    // cached the download response in the session
+    if (!response.link.isEmpty) {
+      session.getUserProperties
+        .put(getSessionName(request.downloadType), response)
+    }
+    response
+  }
+
+  private def getSessionName(downloadType: String): String = {
+    s"${SESSION_CACHED_KEY_NAME}_${downloadType}"
+  }
+
+  private def handleGoogleSheetRequest(
+      resultDownloadRequest: ResultDownloadRequest,
+      result: List[ITuple]
+  ): ResultDownloadResponse = {
+    // create google sheet
+    val sheetService: Sheets = GoogleResource.createSheetService()
+    val sheetId: String =
+      createGoogleSheet(sheetService, resultDownloadRequest.workflowName)
+    if (sheetId == null)
+      return ResultDownloadResponse(
+        resultDownloadRequest.downloadType,
+        "",
+        "Fail to create google sheet"
+      )
+
+    // upload the schema
+    val schema: util.List[AnyRef] = getSchema(result)
+    val schemaContent: util.List[util.List[AnyRef]] = Lists.newArrayList()
+    schemaContent.add(schema)
+    val response: AppendValuesResponse =
+      uploadContent(sheetService, sheetId, schemaContent)
+
+    // allow user to access this sheet in the service account
+    val drive: Drive = GoogleResource.createDriveService()
+    val sharePermission: Permission = new Permission()
+      .setType("anyone")
+      .setRole("reader")
+    drive
+      .permissions()
+      .create(sheetId, sharePermission)
+      .execute()
+
+    // upload the content asynchronously to avoid long waiting on the user side.
+    // may change to thread pool
+    new Thread(() => uploadResult(sheetService, sheetId, result)).start()
```

Comment:
when does this thread join?

Reply:
May I ask the reason to join this thread? The reason to use a thread here is to give the link back to user without waiting the content uploading to finish. The thread here is to upload the content and I think there's no need to wait for it.